lib/SILOptimizer/Utils/PartialApplyCombiner.cpp

```diff
@@ -229,78 +139,80 @@ bool PartialApplyCombiner::processSingleApply(FullApplySite paiAI) {
   // pre-incremented. When we combine the partial_apply and this apply into
   // a new apply we need to retain all of the closure non-address type
   // arguments.
+  auto destroyloc = RegularLocation::getAutoGeneratedLocation();
   auto paramInfo = pai->getSubstCalleeType()->getParameters();
   auto partialApplyArgs = pai->getArguments();
-  // Set of arguments that need to be destroyed after each invocation.
-  SmallVector<SILValue, 8> toBeDestroyedArgs;
   for (unsigned i : indices(partialApplyArgs)) {
-    auto arg = partialApplyArgs[i];
+    SILValue arg = partialApplyArgs[i];
+    if (argToTmpCopy.count(arg))
+      arg = argToTmpCopy.lookup(arg);
 
-    if (!arg->getType().isAddress()) {
+    if (paramInfo[paramInfo.size() - partialApplyArgs.size() + i]
+            .isConsumed()) {
       // Copy the argument as the callee may consume it.
-      arg = builder.emitCopyValueOperation(pai->getLoc(), arg);
-      // For non consumed parameters (e.g. guaranteed), we also need to
-      // insert destroys after each apply instruction that we create.
-      if (!paramInfo[paramInfo.size() - partialApplyArgs.size() + i]
-               .isConsumed())
-        toBeDestroyedArgs.push_back(arg);
+      if (arg->getType().isAddress()) {
+        auto *ASI = builder.createAllocStack(pai->getLoc(), arg->getType());
+        builder.createCopyAddr(pai->getLoc(), arg, ASI, IsTake_t::IsNotTake,
```

Comment:
Thanks for fixing this. We should be copying before each call site. This would be another place where SILGen's ManagedValue/Scope machinery would be useful.